File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.java
class Solution {
    public int minimumEffortPath(int[][] heights) {
        int rows = heights.length;
        int cols = heights[0].length;
        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
        int[][] distance = new int[rows][cols];
        for (int[] row : distance) {
            Arrays.fill(row, Integer.MAX_VALUE);
        }
        distance[0][0] = 0;

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        pq.offer(new int[]{0, 0, 0});
        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int effort = cur[0];
            int row = cur[1];
            int col = cur[2];

            if (row == rows - 1 && col == cols - 1) return effort;

            if (effort > distance[row][col]) continue;

            for (int[] direction : directions) {
                int newRow = row + direction[0];
                int newCol = col + direction[1];

                if (newRow < 0 || newRow >= rows || newCol < 0 || newCol >= cols) continue;

                int heightDiff = Math.abs(heights[newRow][newCol] - heights[row][col]);
                int newEffort = Math.max(effort, heightDiff);

                if (newEffort < distance[newRow][newCol]) {
                    distance[newRow][newCol] = newEffort;
                    pq.offer(new int[]{newEffort, newRow, newCol});
                }
            }
        }

        return 0;
    }
}

File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.kt
import kotlin.math.abs

class Solution {
    fun minimumEffortPath(heights: Array<IntArray>): Int {
        val rows = heights.size
        val cols = heights[0].size
        val directions = arrayOf(
            intArrayOf(-1, 0),
            intArrayOf(1, 0),
            intArrayOf(0, -1),
            intArrayOf(0, 1)
        )
        val distance = Array(rows) { IntArray(cols) { Int.MAX_VALUE } }
        distance[0][0] = 0

        val pq = PriorityQueue<IntArray>(compareBy { it[0] })
        pq.offer(intArrayOf(0, 0, 0))
        while (pq.isNotEmpty()) {
            val (effort, row, col) = pq.poll()

            if (row == rows - 1 && col == cols - 1) return effort

            if (effort > distance[row][col]) continue

            for ((dr, dc) in directions) {
                val newRow = row + dr
                val newCol = col + dc

                if (newRow !in 0 until rows || newCol !in 0 until cols) continue

                val heightDiff = abs(heights[newRow][newCol] - heights[row][col])
                val newEffort = maxOf(effort, heightDiff)

                if (newEffort < distance[newRow][newCol]) {
                    distance[newRow][newCol] = newEffort
                    pq.offer(intArrayOf(newEffort, newRow, newCol))
                }
            }
        }

        return 0
    }
}